Global Styles: Allow references to values in other locations in the tree

scruffian opened the pull request:
## What?
There are many times when a block would like to access properties from the global styles context but is unable to do so. Sometimes that isn't a problem as the CSS cascade takes care of it for us, but in the case where blocks want to apply a setting to a different property, there's no mechanism to do so. This PR proposes that we add a new convention to theme.json which allows one property to set itself to the value of another, for example buttons could do this:

```
{
	"color": {
		"text": { ref: "styles.color.background" }
	}
}
```

Global Styles will convert `{ ref: "styles.color.background" }` to whatever is the value at `styles > color > background` in theme.json.

## Why?
This will allow blocks to be a lot less opinionated in their styles and simply inherit from theme settings.

## How?
The main part of this change is in `get_property_value`, which adds some lines of code to detect whether the value starts with `styles` or `settings`. If it does then we replace it with the result of _wp_array_get. The rest of the changes are necessary to pass $theme_json down to this function, since it's static. This was first attempted in https://github.com/WordPress/gutenberg/pull/39432.

## Testing Instructions
- Switch to emptytheme
- Add these blocks to a post: Button, File, Search
- Check that the colors of these button are the inverse of the colors of the background and text of the site.

## Screenshots or screencast <!-- if applicable -->
<img width="528" alt="Screenshot 2022-06-13 at 17 19 37" src="https://user-images.githubusercontent.com/275961/173399179-54adcb5e-3f09-4a13-a03b-02379c0ac30f.png">


## Note
This change has a few side effects that we'll need to be careful with
1. The default colors for theme.json are now fixed, so they won't fall back to the browser defaults.
2. Search block buttons will now match other buttons in visual appearance, which they didn't before.

cc @WordPress/block-themers

scruffian commented on lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php:
```diff
@@ -301,4 +301,117 @@ protected function get_block_classes( $style_nodes ) {
+	 * @param array $properties Properties metadata.
+	 * @return array Returns the modified $declarations.
+	 */
+	protected static function compute_style_properties( $styles, $settings = array(), $properties = null, $theme_json = null ) {
```
No change to this function except that we're passing in `theme_json`.

scruffian commented on lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php:
```diff
@@ -301,4 +301,117 @@ protected function get_block_classes( $style_nodes ) {
+		}
+
+		foreach ( $properties as $css_property => $value_path ) {
+			$value = static::get_property_value( $styles, $value_path, $theme_json );
```
And then passing `$theme_json` through here...

scruffian commented on lib/experimental/class-wp-theme-json-resolver-gutenberg.php:
```diff
@@ -177,5 +177,21 @@ public static function get_merged_data( $origin = 'custom' ) {
+	 *
+	 * @return WP_Theme_JSON_Gutenberg Entity that holds core data.
+	 */
+	public static function get_core_data() {
```
This is just here so that we grab the updated theme.json file.

scruffian commented on lib/experimental/theme.json:
```diff
@@ -0,0 +1,256 @@
+			}
+		}
+	},
+	"styles": {
```
This file is the same apart from the `styles` section which now has settings for colors and elements.

andrewserong commented on lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php:
```diff
@@ -301,4 +301,117 @@ protected function get_block_classes( $style_nodes ) {
+		if ( $has_styles_prefix || $has_settings_prefix ) {
+			// Get the path of the style/setting.
+			$value_path = explode( '.', $value );
+			$new_value = _wp_array_get( $theme_json, $value_path );
```
What if the value that's been retrieved is a value that begins with the `var:` prefix?

Should we move this block that does the look up to the beginning of this function, so that the retrieved value can still be parsed via the `var:` check?

As this gets more complex, it could be good for us to write up a few test cases to get some coverage 🙂

adamziel replied:
This! Otherwise we'll quickly discover that some actual values are expressed using the exact same strings as these references. I would even go a step further and use an object like `{ source: "styles.color.background" }`  or so, otherwise we're effectively packing a data structure into a string.

scruffian replied:
Good idea, done in 7b9f7ae6da

andrewserong commented on lib/experimental/theme.json:
```diff
@@ -0,0 +1,257 @@
+	},
+	"styles": {
+		"color": {
+			"background": "#fff",
```
Is this potentially a large change? Some classic themes incrementally adopt `theme.json` and don't necessarily set a background or text color, so imposing one in the base `theme.json` could have fairly big side-effects?

What do you think about separating this feature out into two PRs — one that implements the settings/style lookup feature, and one that proposes setting the base / default colors in theme.json, and the button to inherit that color?

scruffian replied:
Yeah I think that's a good idea, those changes are mostly there to help with testing!

scruffian replied:
Done

andrewserong commented on lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php:
```diff
@@ -301,4 +301,117 @@ protected function get_block_classes( $style_nodes ) {
+			$new_value = _wp_array_get( $theme_json, $value_path );
+			// Only use the new value if we find anything.
+			if ( ! empty( $new_value ) ) {
+				$value = $new_value;
```
Should we add some validation here? What if the setting that was looked up is a boolean, null, or array?

scruffian replied:
Good idea. Done in 444f1753d9.

andrewserong reviewed (commented):
Very cool idea @scruffian! I like the idea of the styles/settings lookup, and I can imagine it being quite powerful, and it seems a bit more direct than some cases of trying to implement this via CSS variables instead like in your earlier PR (https://github.com/WordPress/gutenberg/pull/39432)

I've just left a couple of comments — while I think the feature itself is a great idea, I was wondering if it's worth breaking up the PR slightly to propose the new feature in one PR, and propose setting the default colors in another PR?

andrewserong commented on lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php:
```diff
@@ -301,4 +301,117 @@ protected function get_block_classes( $style_nodes ) {
+		$styles_prefix       = 'styles';
+		$styles_prefix_len   = strlen( $styles_prefix );
+		$has_styles_prefix   = 0 === strncmp( $value, $styles_prefix, $styles_prefix_len );
+		$settings_prefix     = 'settings';
```
What's the use-case for `settings` lookups? From the looks of it, currently, most settings aren't direct values that can be easily used in styles (either the settings are non-string values like booleans, or if they're presets, they're in an array which can't be access easily via dot notation).

scruffian replied:
I think you're right, lets keep it to `styles` for now.

scruffian replied:
Done in 444f1753d9.

draganescu commented on lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php:
```diff
@@ -301,4 +301,117 @@ protected function get_block_classes( $style_nodes ) {
+
+			// Look up protected properties, keyed by value path.
+			// Skip protected properties that are explicitly set to `null`.
+			//var_dump( $value_path );
```
Just flagging the commented `var_dump`.

scruffian replied:
Fixed, thanks!

draganescu commented:
Cool stuff 👍🏻 I think it's awesome to lean in the power of our chosen format! 

I wonder if this will introduce a lot of complexity in themes as it is so inviting to do this. Will it lower the expectation for defined palettes for instance? E.g. instead of "primary-color"  just use "styles.color.background" ...

MaggieCabrera commented:
> Cool stuff 👍🏻 I think it's awesome to lean in the power of our chosen format!
> 
> I wonder if this will introduce a lot of complexity in themes as it is so inviting to do this. Will it lower the expectation for defined palettes for instance? E.g. instead of "primary-color" just use "styles.color.background" ...

I think this is great to have. The defined palette is important just on its own because it's user-facing. And in your example, the background color is going to be defined as primary in any case, this approach just allows a different way of referencing colors and adds an extra tool for plugins to be styled in line with the rest of the theme.

scruffian commented:
I've addressed the feedback here. If y'all are happy with this approach, I'll remove the changes to theme.json so that this is a lot simpler. I'll leave them there for now to make it easy to test :)

adamziel commented:
cc @ramonjd

adamziel commented on lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php:
```diff
@@ -301,4 +301,108 @@ protected function get_block_classes( $style_nodes ) {
+	protected static function get_property_value( $styles, $path, $theme_json = null ) {
+		$value = _wp_array_get( $styles, $path, '' );
+
+		if( is_array( $value ) && array_key_exists( 'source', $value ) ) {
```
I'd add a short inline comment to explain what it does

ramonjd replied:
This is probably a nothing burger, but what do you think about `"ref"` as opposed to `"source"` to reference a static JSON value?

It only occurred to me since JSON schema uses `"$ref"` for internal pointers.

Not a biggie.

scruffian replied:
WFM

scruffian replied:
Done

adamziel commented on lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php:
```diff
@@ -301,4 +301,108 @@ protected function get_block_classes( $style_nodes ) {
+			$value_path = explode( '.', $value['source'] );
+			$source_value = _wp_array_get( $theme_json, $value_path );
+			// Only use the source value if we find anything.
+			if ( ! empty( $source_value ) && is_string( $source_value ) ) {
```
If $source_value is empty, would this return just `{ source: "..." }`? Is this intended? Also, could an empty value be actually valid and desired?

scruffian replied:
My thinking is that if nothing is found then we should be non-destructive. This could be useful in the case where multiple theme.json files are combined - we might not have a value on the first processing but we may do later...

adamziel replied:
Could this lead to a situation where the final value after all the processing ends up being `{ source: "..." }`?

scruffian replied:
yes, do you think that could be a problem?

adamziel replied:
Not sure, depends on how it gets rendered as CSS –  I haven't tried running it yet 🤔

scruffian replied:
If the value is an array it get skipped inside `compute_style_properties` (line 348).

jffng commented:
🙌 

I tried the following using [Curator](https://github.com/Automattic/themes/tree/trunk/curator) and it also worked, the background color of button elements was set to the same value as the button block. I also tested changing the button block background color via the GS panel (blocks > button > color > background), and the button elements still matched.

```
	"styles": {
		"elements": {
			"button": {
				"color": {
					"background": { 
						"source": "styles.blocks.core/button.color.background"
					}
				}
			}
		}
	}
```

This feels pretty powerful. It creates possibilities for themes to make some strange connections / "links" in addition to very sensible ones. 

How would this work if I wanted to access a color from `settings` > `color` > `palette`?

scruffian commented:
> How would this work if I wanted to access a color from settings > color > palette?

We already have a mechanism for doing that, using vars (`var:preset|color|secondary`)

ramonjd commented:
I like this idea. Thanks @scruffian 

It's testing pretty well for me so far. I was toying with some edge cases, and I wasn't sure if they were important or not.

One was: should a ref to another ref be able to resolve (I can't see that it does)? E.g., see "margin" here: 

```json
		"spacing": {
			"blockGap": "200px",
			"padding": {
				"source": "styles.spacing.blockGap"
			},
			"margin": {
				"source": "styles.spacing.padding"
			}
		},
```

Just asking. I could imagine that would get pretty recursive.

draganescu commented:
@ramonjd 's example is my main fear with this enhancement, it's a tool for complexity creation, which is solving the same problem that css variables solves but has the potential to bring over a different flavor of callback hell.

scruffian commented:
> One was: should a ref to another ref be able to resolve (I can't see that it does)? E.g., see "margin" here:

Yeah at the moment it doesn't work. I feel like this is probably the best option to avoid any kind of recursive loops etc. If you want to use a property which references another property then you'll have to update the reference to the parent property.

adamziel commented:
> Yeah at the moment it doesn't work. I feel like this is probably the best option to avoid any kind of recursive loops etc. If you want to use a property which references another property then you'll have to update the reference to the parent property.

Adding a `_doing_it_wrong` call or a similar warning in case of recursion could be really helpful to future developers debugging their code.

scruffian committed:
Elements: Add an API make it easier to get class names (#41753)

* Elements: Add an api make it easier to get class names

* Add some PHP unit tests

* Add a unit tests for elements

* lint fix

* fix PHP tests

* Update lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php

Co-authored-by: George Mamadashvili <georgemamadashvili@gmail.com>

* Add @since

* update form.js

* fix PHP test

Co-authored-by: George Mamadashvili <georgemamadashvili@gmail.com>

Theme JSON: Add dynamic properties

Theme JSON: Make it possible to use styles and settings from elsewhere in the tree

also make button text dynamic

fix type

remove unconnected changes

move the value replacement earlier and add validation

use an object rather than a string to fetch other values

adding a doing it wrong message

remove color changes for buttons

add an explanatory comment

change source to ref

Add unit tests

Update lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php

Co-authored-by: Adam Zielinski <adam@adamziel.com>

lint fixes

fix fixtures

remove incorrect change

lint fixes

scruffian commented:
> Adding a _doing_it_wrong call or a similar warning in case of recursion could be really helpful to future developers debugging their code.

Done!

scruffian commented:
I've made a few changes here:

1. Changed `source` to `ref`
2. Added a _doing_it_wrong notice for the case where someone might reference another dynamic property
3. Removed the new theme.json file and the changes to buttons.

Ready for another review...

andrewserong commented:
Thanks for the updates @scruffian! 

> Changed source to ref

It looks like the PR still refers to `source`, is there a missing commit?

Also, since this introduces some potentially complex logic, would it be worth adding a couple of PHP tests with example `theme.json` structures provided to ensure the lookup and `_doing_it_wrong` calls are behaving as expected? With the style engine work, at some stage, this logic might need to be migrated elsewhere, so it'd be helpful to have some working tests to ensure we've got a stable baseline for the behaviour.

scruffian commented:
> It looks like the PR still refers to source, is there a missing commit?

Ooops sorry yes, try now!

Good idea, I'll add some tests...

adamziel commented on lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php:
```diff
@@ -317,4 +317,117 @@ protected function get_block_classes( $style_nodes ) {
+			if ( is_array( $ref_value ) && array_key_exists( 'ref', $ref_value ) ) {
+				$path_string = json_encode( $path );
+				$ref_value_string = json_encode( $ref_value );
+				_doing_it_wrong( 'get_property_value', "Your theme.json file uses a dynamic value (${ref_value_string}) for the path at ${path_string}. Please update this to point to the value at ${ref_value_string}.", '6.1.0' );
```
```suggestion
				_doing_it_wrong( 'get_property_value', "Your theme.json file uses a dynamic value (${ref_value_string}) for the path at ${path_string}. However, the value at ${path_string} is also a dynamic value (pointing to ${ref_value['ref']}) and pointing to another dynamic value is not supported. Please update ${path_string} to point directly to ${ref_value['ref']}.", '6.1.0' );
```

scruffian commented:
I added some unit tests in [b607e0f](https://github.com/WordPress/gutenberg/pull/41696/commits/b607e0f082cff144088fc0bc50dfe7c57446db6c)

scruffian committed:
fix rebase

oandregal commented:
While getting familiar with how this works I've found a bug (this doesn't work in the site editor) https://github.com/WordPress/gutenberg/issues/42890

scruffian commented:
Thanks @oandregal I noticed that too: https://github.com/WordPress/gutenberg/issues/42884

It's on my list of things to look at this week.

adamziel commented:
For posterity, [YAML supports these kind of aliases out of the box](https://yaml.org/spec/1.2.2/#71-alias-nodes). It also supports comments.

scruffian commented:
Dev Note draft:

## What
Theme designs often require consistency in the styles applied to blocks. Often this can be achieved by setting Global Styles properties which are inherited by blocks using the inheritance of the CSS cascade. In some cases blocks want to apply settings from Global Styles to a different property of the block - for example the button element might want to use the global text color for its background color and the global background for it's text color.

## Why
To solve this problem we need to allow themes to share Global Styles settings with blocks. This will make it easier for users to update these shared properties across all their blocks. Again take the example of a button element that wants to use the text color for its background. If the button element is set up as in the example above, then when a user edits the global text color, the background color of their buttons will also be updated.

## How
To achieve this, we have added new `ref` property to theme.json which allows one property to set itself to the value of another, for example we could define this for button elements:
```
		"elements": {
			"button": {
				"color": {
					"background": {
						"ref": "styles.color.text"
					},
					"text": {
						"ref": "styles.color.background"
					}
				}
			}
		}
```

Global Styles will convert { ref: "styles.color.background" } the value at: `styles > color > background` in theme.json.


## Limitations
It is currently only possible to use `ref` to get one value from theme.json; a `ref` cannot point to another `ref`. This also prevents circular `ref`s from causing problems.
